quil-rs/src/parser/common.rs

```diff
@@ -195,6 +195,60 @@ pub(crate) fn parse_permutation<'a>(input: ParserInput<'a>) -> InternalParserRes
     )(input)
 }
 
+/// Parse a collection of [`crate::instruction::PauliWord`]s. These are a special kind of a
+/// [`Token::Identifier`] used in Pauli sum `DEFGATE` specifications where each identifier
+/// is a single string made up of one or more `I`, `X`, `Y`, `Z` characters (e.g. `X`, `YY`),
+/// each representing a Pauli word.
+fn parse_pauli_words<'a>(input: ParserInput<'a>) -> InternalParserResult<'a, Vec<PauliWord>> {
+    map_res(token!(Identifier(v)), |words| {
+        let mut pauli_words: Vec<PauliWord> = Vec::new();
+        for word in words.split("") {
+            if word.is_empty() {
+                // split("") separates every character, including empty spaces at the beginning and
+                // end of string.
+                continue;
+            }
+            pauli_words.push(PauliWord::from_str(word).map_err(|_| {
+                InternalParseError::from_kind(
+                    input,
+                    ParserErrorKind::ExpectedCharacter {
+                        expected: "(I,X,Y,Z)".to_string(),
+                        actual: word.to_string(),
+                    },
+                )
+            })?)
+        }
+        Ok(pauli_words)
+    })(input)
+}
+
+/// Parse Pauli sum representation of a `DEFGATE` specification.
+pub(crate) fn parse_pauli_sum<'a>(
+    input: ParserInput<'a>,
+) -> InternalParserResult<'a, Vec<PauliTerm>> {
+    preceded(
+        token!(NewLine),
+        separated_list1(
+            token!(NewLine),
+            preceded(
+                token!(Indentation),
+                map(
+                    tuple((
+                        parse_pauli_words,
+                        delimited(token!(LParenthesis), parse_expression, token!(RParenthesis)),
+                        many1(token!(Identifier(i))),
+                    )),
+                    |(words, expression, arguments)| PauliTerm {
+                        words,
+                        expression,
+                        arguments,
+                    },
```

Comment:
Per the Quil spec, the length of `words` and `arguments` should be the same. Is there a good pattern to enforce this kind of rule in the parser?

If not, maybe this a case where we could use a `new` constructor to perform additional validation in the parser?

Reply:
Maybe [verify](https://docs.rs/nom/latest/nom/combinator/fn.verify.html)?

Reply:
The catch will be if the error types are awkward to mesh together, but you had that pattern in use already on a previous MR so I assume it works out alright.

Reply:
I think using the `PauliTerm::new` here makes sense 👍

Reply:
Done!